File: genetico/caixeiro-viajante/js/logica/genetico.js
import { criarGeradorAleatorio, embaralhar, sortearInteiro } from './aleatorio.js';
import { comprimentoDaRota, criarMatrizDeDistancias } from './distancias.js';

export const QUANTIDADE_MINIMA_DE_PONTOS = 4;

export const CONFIGURACAO_PADRAO = {
  tamanhoDaPopulacao: 120,
  taxaDeCruzamento: 0.9,
  taxaDeMutacao: 0.5,
  quantidadeDeElite: 2,
  tamanhoDoTorneio: 3,
  geracoesSemMelhoraParaParar: 800,
};

export function criarIndividuo(rota, matrizDeDistancias) {
  const distancia = comprimentoDaRota(rota, matrizDeDistancias);
  return { rota, distancia, aptidao: 1 / distancia };
}

function ordenarPorDistancia(populacao) {
  return populacao.sort((individuoA, individuoB) => individuoA.distancia - individuoB.distancia);
}

export function criarPopulacaoInicial(quantidadeDePontos, tamanhoDaPopulacao, matrizDeDistancias, geradorAleatorio) {
  const rotaBase = Array.from({ length: quantidadeDePontos }, (_, indice) => indice);
  const populacao = Array.from({ length: tamanhoDaPopulacao }, () =>
    criarIndividuo(embaralhar(rotaBase, geradorAleatorio), matrizDeDistancias),
  );
  return ordenarPorDistancia(populacao);
}

export function selecionarPorTorneio(populacao, tamanhoDoTorneio, geradorAleatorio) {
  const participantes = Array.from(
    { length: tamanhoDoTorneio },
    () => populacao[sortearInteiro(geradorAleatorio, 0, populacao.length - 1)],
  );
  const vencedor = participantes.reduce((melhor, participante) =>
    participante.distancia < melhor.distancia ? participante : melhor,
  );
  return { participantes, vencedor };
}

function sortearTrecho(tamanho, geradorAleatorio) {
  const primeiro = sortearInteiro(geradorAleatorio, 0, tamanho - 1);
  const segundo = sortearInteiro(geradorAleatorio, 0, tamanho - 1);
  return { inicio: Math.min(primeiro, segundo), fim: Math.max(primeiro, segundo) };
}

export function cruzarComOrderCrossover(rotaDoPaiA, rotaDoPaiB, geradorAleatorio) {
  const tamanho = rotaDoPaiA.length;
  const { inicio, fim } = sortearTrecho(tamanho, geradorAleatorio);
  const rotaDoFilho = new Array(tamanho).fill(-1);
  const pontosHerdadosDoPaiA = new Set();

  for (let posicao = inicio; posicao <= fim; posicao++) {
    rotaDoFilho[posicao] = rotaDoPaiA[posicao];
    pontosHerdadosDoPaiA.add(rotaDoPaiA[posicao]);
  }

  let posicaoParaPreencher = (fim + 1) % tamanho;
  for (let deslocamento = 0; deslocamento < tamanho; deslocamento++) {
    const pontoDoPaiB = rotaDoPaiB[(fim + 1 + deslocamento) % tamanho];
    if (!pontosHerdadosDoPaiA.has(pontoDoPaiB)) {
      rotaDoFilho[posicaoParaPreencher] = pontoDoPaiB;
      posicaoParaPreencher = (posicaoParaPreencher + 1) % tamanho;
    }
  }
  return { rota: rotaDoFilho, inicioDoTrecho: inicio, fimDoTrecho: fim };
}

export function mutarPorInversao(rota, geradorAleatorio) {
  const { inicio, fim } = sortearTrecho(rota.length, geradorAleatorio);
  const rotaMutada = [...rota.slice(0, inicio), ...rota.slice(inicio, fim + 1).reverse(), ...rota.slice(fim + 1)];
  return { rota: rotaMutada, inicioDoTrecho: inicio, fimDoTrecho: fim };
}

function calcularEstatisticas(geracao, populacao) {
  const somaDasDistancias = populacao.reduce((soma, individuo) => soma + individuo.distancia, 0);
  return {
    geracao,
    melhor: populacao[0].distancia,
    media: somaDasDistancias / populacao.length,
    pior: populacao.at(-1).distancia,
  };
}

export function criarAlgoritmoGenetico(pontos, configuracao, semente) {
  if (pontos.length < QUANTIDADE_MINIMA_DE_PONTOS) {
    throw new Error(`São necessários pelo menos ${QUANTIDADE_MINIMA_DE_PONTOS} pontos.`);
  }
  const geradorAleatorio = criarGeradorAleatorio(semente);
  const matrizDeDistancias = criarMatrizDeDistancias(pontos);
  const populacao = criarPopulacaoInicial(
    pontos.length,
    configuracao.tamanhoDaPopulacao,
    matrizDeDistancias,
    geradorAleatorio,
  );
  const estatisticasIniciais = calcularEstatisticas(0, populacao);
  return {
    pontos,
    matrizDeDistancias,
    geradorAleatorio,
    populacao,
    geracao: 0,
    historico: [estatisticasIniciais],
    geracaoDaUltimaMelhora: 0,
    individuosAvaliados: populacao.length,
    exemploDeReproducao: null,
  };
}

function gerarFilho(algoritmo, configuracao) {
  const { populacao, geradorAleatorio, matrizDeDistancias } = algoritmo;
  const torneioDoPaiA = selecionarPorTorneio(populacao, configuracao.tamanhoDoTorneio, geradorAleatorio);
  const torneioDoPaiB = selecionarPorTorneio(populacao, configuracao.tamanhoDoTorneio, geradorAleatorio);

  const cruzamento =
    geradorAleatorio() < configuracao.taxaDeCruzamento
      ? cruzarComOrderCrossover(torneioDoPaiA.vencedor.rota, torneioDoPaiB.vencedor.rota, geradorAleatorio)
      : null;
  const rotaAntesDaMutacao = cruzamento ? cruzamento.rota : [...torneioDoPaiA.vencedor.rota];

  const mutacao = geradorAleatorio() < configuracao.taxaDeMutacao ? mutarPorInversao(rotaAntesDaMutacao, geradorAleatorio) : null;
  const filho = criarIndividuo(mutacao ? mutacao.rota : rotaAntesDaMutacao, matrizDeDistancias);

  return { filho, exemplo: { torneioDoPaiA, torneioDoPaiB, cruzamento, rotaAntesDaMutacao, mutacao, filho } };
}

export function evoluirUmaGeracao(algoritmo, configuracao) {
  const quantidadeDeElite = Math.min(configuracao.quantidadeDeElite, configuracao.tamanhoDaPopulacao);
  const novaPopulacao = algoritmo.populacao.slice(0, quantidadeDeElite);
  let exemploDeReproducao = null;

  while (novaPopulacao.length < configuracao.tamanhoDaPopulacao) {
    const { filho, exemplo } = gerarFilho(algoritmo, configuracao);
    exemploDeReproducao ??= exemplo;
    novaPopulacao.push(filho);
  }

  const melhorDistanciaAnterior = algoritmo.populacao[0].distancia;
  algoritmo.populacao = ordenarPorDistancia(novaPopulacao);
  algoritmo.geracao += 1;
  algoritmo.individuosAvaliados += configuracao.tamanhoDaPopulacao - quantidadeDeElite;
  algoritmo.exemploDeReproducao = exemploDeReproducao;
  algoritmo.historico.push(calcularEstatisticas(algoritmo.geracao, algoritmo.populacao));
  if (algoritmo.populacao[0].distancia < melhorDistanciaAnterior - 1e-9) {
    algoritmo.geracaoDaUltimaMelhora = algoritmo.geracao;
  }
  return algoritmo;
}

export function algoritmoConvergiu(algoritmo, configuracao) {
  return algoritmo.geracao - algoritmo.geracaoDaUltimaMelhora >= configuracao.geracoesSemMelhoraParaParar;
}

export function melhorIndividuo(algoritmo) {
  return algoritmo.populacao[0];
}

export function contarRotasDistintas(populacao) {
  return new Set(populacao.map((individuo) => individuo.rota.join(','))).size;
}

File: genetico/logica/genetico.py
import random
from dataclasses import dataclass, field

from .cidades import Ponto
from .distancias import comprimento_da_rota, criar_matriz_de_distancias

QUANTIDADE_MINIMA_DE_PONTOS = 4


@dataclass(frozen=True)
class ConfiguracaoDoAlgoritmo:
    tamanho_da_populacao: int = 120
    taxa_de_cruzamento: float = 0.9
    taxa_de_mutacao: float = 0.5
    quantidade_de_elite: int = 2
    tamanho_do_torneio: int = 3
    geracoes_sem_melhora_para_parar: int = 800


@dataclass(frozen=True)
class Individuo:
    rota: tuple[int, ...]
    distancia: float

    @property
    def aptidao(self) -> float:
        return 1 / self.distancia


@dataclass(frozen=True)
class ResultadoDoTorneio:
    participantes: tuple[Individuo, ...]
    vencedor: Individuo


@dataclass(frozen=True)
class ResultadoDoCruzamento:
    rota: tuple[int, ...]
    inicio_do_trecho: int
    fim_do_trecho: int


@dataclass(frozen=True)
class ResultadoDaMutacao:
    rota: tuple[int, ...]
    inicio_do_trecho: int
    fim_do_trecho: int


@dataclass(frozen=True)
class ExemploDeReproducao:
    torneio_do_pai_a: ResultadoDoTorneio
    torneio_do_pai_b: ResultadoDoTorneio
    cruzamento: ResultadoDoCruzamento | None
    rota_antes_da_mutacao: tuple[int, ...]
    mutacao: ResultadoDaMutacao | None
    filho: Individuo


@dataclass(frozen=True)
class EstatisticasDaGeracao:
    geracao: int
    melhor: float
    media: float
    pior: float


@dataclass
class AlgoritmoGenetico:
    pontos: list[Ponto]
    matriz_de_distancias: list[list[float]]
    gerador_aleatorio: random.Random
    populacao: list[Individuo]
    geracao: int = 0
    historico: list[EstatisticasDaGeracao] = field(default_factory=list)
    geracao_da_ultima_melhora: int = 0
    individuos_avaliados: int = 0
    exemplo_de_reproducao: ExemploDeReproducao | None = None


def criar_individuo(rota: tuple[int, ...], matriz_de_distancias: list[list[float]]) -> Individuo:
    return Individuo(rota=rota, distancia=comprimento_da_rota(rota, matriz_de_distancias))


def _ordenar_por_distancia(populacao: list[Individuo]) -> list[Individuo]:
    return sorted(populacao, key=lambda individuo: individuo.distancia)


def criar_populacao_inicial(
    quantidade_de_pontos: int,
    tamanho_da_populacao: int,
    matriz_de_distancias: list[list[float]],
    gerador_aleatorio: random.Random,
) -> list[Individuo]:
    rota_base = list(range(quantidade_de_pontos))
    populacao = [
        criar_individuo(tuple(gerador_aleatorio.sample(rota_base, quantidade_de_pontos)), matriz_de_distancias)
        for _ in range(tamanho_da_populacao)
    ]
    return _ordenar_por_distancia(populacao)


def selecionar_por_torneio(
    populacao: list[Individuo], tamanho_do_torneio: int, gerador_aleatorio: random.Random
) -> ResultadoDoTorneio:
    participantes = tuple(gerador_aleatorio.choice(populacao) for _ in range(tamanho_do_torneio))
    vencedor = min(participantes, key=lambda participante: participante.distancia)
    return ResultadoDoTorneio(participantes=participantes, vencedor=vencedor)


def _sortear_trecho(tamanho: int, gerador_aleatorio: random.Random) -> tuple[int, int]:
    primeiro = gerador_aleatorio.randint(0, tamanho - 1)
    segundo = gerador_aleatorio.randint(0, tamanho - 1)
    return min(primeiro, segundo), max(primeiro, segundo)


def cruzar_com_order_crossover(
    rota_do_pai_a: tuple[int, ...], rota_do_pai_b: tuple[int, ...], gerador_aleatorio: random.Random
) -> ResultadoDoCruzamento:
    tamanho = len(rota_do_pai_a)
    inicio, fim = _sortear_trecho(tamanho, gerador_aleatorio)
    rota_do_filho = [-1] * tamanho
    rota_do_filho[inicio : fim + 1] = rota_do_pai_a[inicio : fim + 1]
    pontos_herdados_do_pai_a = set(rota_do_pai_a[inicio : fim + 1])

    posicao_para_preencher = (fim + 1) % tamanho
    for deslocamento in range(tamanho):
        ponto_do_pai_b = rota_do_pai_b[(fim + 1 + deslocamento) % tamanho]
        if ponto_do_pai_b not in pontos_herdados_do_pai_a:
            rota_do_filho[posicao_para_preencher] = ponto_do_pai_b
            posicao_para_preencher = (posicao_para_preencher + 1) % tamanho
    return ResultadoDoCruzamento(rota=tuple(rota_do_filho), inicio_do_trecho=inicio, fim_do_trecho=fim)


def mutar_por_inversao(rota: tuple[int, ...], gerador_aleatorio: random.Random) -> ResultadoDaMutacao:
    inicio, fim = _sortear_trecho(len(rota), gerador_aleatorio)
    rota_mutada = rota[:inicio] + rota[inicio : fim + 1][::-1] + rota[fim + 1 :]
    return ResultadoDaMutacao(rota=rota_mutada, inicio_do_trecho=inicio, fim_do_trecho=fim)


def _calcular_estatisticas(geracao: int, populacao: list[Individuo]) -> EstatisticasDaGeracao:
    distancias = [individuo.distancia for individuo in populacao]
    return EstatisticasDaGeracao(
        geracao=geracao,
        melhor=distancias[0],
        media=sum(distancias) / len(distancias),
        pior=distancias[-1],
    )


def criar_algoritmo_genetico(
    pontos: list[Ponto], configuracao: ConfiguracaoDoAlgoritmo, semente: int
) -> AlgoritmoGenetico:
    if len(pontos) < QUANTIDADE_MINIMA_DE_PONTOS:
        raise ValueError(f"São necessários pelo menos {QUANTIDADE_MINIMA_DE_PONTOS} pontos.")
    gerador_aleatorio = random.Random(semente)
    matriz_de_distancias = criar_matriz_de_distancias(pontos)
    populacao = criar_populacao_inicial(
        len(pontos), configuracao.tamanho_da_populacao, matriz_de_distancias, gerador_aleatorio
    )
    return AlgoritmoGenetico(
        pontos=pontos,
        matriz_de_distancias=matriz_de_distancias,
        gerador_aleatorio=gerador_aleatorio,
        populacao=populacao,
        historico=[_calcular_estatisticas(0, populacao)],
        individuos_avaliados=len(populacao),
    )


def _gerar_filho(algoritmo: AlgoritmoGenetico, configuracao: ConfiguracaoDoAlgoritmo) -> ExemploDeReproducao:
    gerador_aleatorio = algoritmo.gerador_aleatorio
    torneio_do_pai_a = selecionar_por_torneio(algoritmo.populacao, configuracao.tamanho_do_torneio, gerador_aleatorio)
    torneio_do_pai_b = selecionar_por_torneio(algoritmo.populacao, configuracao.tamanho_do_torneio, gerador_aleatorio)

    cruzamento = None
    if gerador_aleatorio.random() < configuracao.taxa_de_cruzamento:
        cruzamento = cruzar_com_order_crossover(
            torneio_do_pai_a.vencedor.rota, torneio_do_pai_b.vencedor.rota, gerador_aleatorio
        )
    rota_antes_da_mutacao = cruzamento.rota if cruzamento else torneio_do_pai_a.vencedor.rota

    mutacao = None
    if gerador_aleatorio.random() < configuracao.taxa_de_mutacao:
        mutacao = mutar_por_inversao(rota_antes_da_mutacao, gerador_aleatorio)
    rota_final = mutacao.rota if mutacao else rota_antes_da_mutacao

    return ExemploDeReproducao(
        torneio_do_pai_a=torneio_do_pai_a,
        torneio_do_pai_b=torneio_do_pai_b,
        cruzamento=cruzamento,
        rota_antes_da_mutacao=rota_antes_da_mutacao,
        mutacao=mutacao,
        filho=criar_individuo(rota_final, algoritmo.matriz_de_distancias),
    )


def evoluir_uma_geracao(algoritmo: AlgoritmoGenetico, configuracao: ConfiguracaoDoAlgoritmo) -> None:
    quantidade_de_elite = min(configuracao.quantidade_de_elite, configuracao.tamanho_da_populacao)
    nova_populacao = algoritmo.populacao[:quantidade_de_elite]
    primeiro_exemplo = None

    while len(nova_populacao) < configuracao.tamanho_da_populacao:
        exemplo = _gerar_filho(algoritmo, configuracao)
        primeiro_exemplo = primeiro_exemplo or exemplo
        nova_populacao.append(exemplo.filho)

    melhor_distancia_anterior = algoritmo.populacao[0].distancia
    algoritmo.populacao = _ordenar_por_distancia(nova_populacao)
    algoritmo.geracao += 1
    algoritmo.individuos_avaliados += configuracao.tamanho_da_populacao - quantidade_de_elite
    algoritmo.exemplo_de_reproducao = primeiro_exemplo
    algoritmo.historico.append(_calcular_estatisticas(algoritmo.geracao, algoritmo.populacao))
    if algoritmo.populacao[0].distancia < melhor_distancia_anterior - 1e-9:
        algoritmo.geracao_da_ultima_melhora = algoritmo.geracao


def algoritmo_convergiu(algoritmo: AlgoritmoGenetico, configuracao: ConfiguracaoDoAlgoritmo) -> bool:
    return algoritmo.geracao - algoritmo.geracao_da_ultima_melhora >= configuracao.geracoes_sem_melhora_para_parar


def melhor_individuo(algoritmo: AlgoritmoGenetico) -> Individuo:
    return algoritmo.populacao[0]


def contar_rotas_distintas(populacao: list[Individuo]) -> int:
    return len({individuo.rota for individuo in populacao})
